src/caches/beliefpropagationcache.jl

```diff
@@ -0,0 +1,221 @@
+default_mt_constructor(inds_e) = ITensor[denseblocks(delta(inds_e))]
+default_mt_storage(ptn::PartitionedGraph) = Dict()
+function default_contractor(contract_list::Vector{ITensor}; kwargs...)
+  return contract_exact(contract_list; kwargs...)
+end
+default_contractor_kwargs() = (; normalize=true, contraction_sequence_alg="optimal")
+@traitfn default_bp_niters(g::::(!IsDirected)) = is_tree(g) ? 1 : nothing
+@traitfn function default_bp_niters(g::::IsDirected)
+  return default_bp_niters(undirected_graph(underlying_graph(g)))
+end
+function message_tensor_diff(mt_a::Vector{ITensor}, mt_b::Vector{ITensor})
+  LHS, RHS = contract(mt_a), contract(mt_b)
+  return 0.5 *
+         norm((denseblocks(LHS) / sum(diag(LHS))) - (denseblocks(RHS) / sum(diag(RHS))))
+end
+
+struct BeliefPropagationCache{PTN,MTS}
+  partitioned_itensornetwork::PTN
+  message_tensors::MTS
+end
```

Comment:
After looking at the code below that uses what's currently called the `default_mt_constructor` (would be good to discuss that name), I think it could make sense to include that in the `BeliefPropagationCache`, since I think it is pretty fundamental to how the cache works (i.e. you need to know how to construct the message tensors if one doesn't exist, so in many ways it is tied to the `message_tensors`).

In terms of the name for that field, I don't like using the shorthand `mt` in `default_mt_constructor` because I think it will be easy to forget what that means (mostly I try to avoid shorthands, unless the name becomes really long without a shorthand or the shorthand is very standard, like "MPS" or "BP"). An alternative is `default_message_tensor`, one issue I see with that is it doesn't indicate it is a function constructing the message tensor, but maybe that is ok and it should be obvious it must be a function.